### PR TITLE
Add schema v2 PaRoutes adapter

### PR DESCRIPTION
**Pull request by anmorgunov**
**Why**
Schema v2 has the route kernel, but the adapt slice needed a v2 adapter boundary for real PaRoutes payloads. PaRoutes extraction should operate on the full target-keyed raw artifact, while casting remains route-local.

**What Changed**
- Added v2 adapter protocol primitives and `RawRouteEntry`.
- Added v2 `Target`, `Task`, `Candidate`, and `FailureRecord` models needed by the adapt path.
- Added a PaRoutes adapter that emits v2 `Route` objects directly.
- Added a reusable v2 adapter contract suite for `iter_raw_routes` and `cast`.
- Added PaRoutes contract and regression coverage, including `tests/testing_data/paroutes.json.gz` and stable route signatures.

**Risk**
The key boundary is `raw PaRoutes payload -> RawRouteEntry -> Route`. The adapter rejects malformed raw schemas, target mismatches, cycles, mixed patents, empty reactions, and strict invalid SMILES with typed error codes. Prune mode drops invalid branches without emitting empty reactions.

**Verification**
- `uv run pytest`
- `uv run pytest tests/v2 --cov=retrocast.v2.adapters.paroutes --cov=retrocast.v2.models.candidates --cov-report=term-missing`
- `uv run ruff check src/retrocast/v2/adapters/paroutes.py tests/v2/adapters/base.py tests/v2/adapters/test_paroutes.py tests/v2/models/test_candidates.py`

**Review**
Self-reviewed against `.practices/docs/common/pr-review.md`; no red/yellow findings remain.

## Summary by CodeRabbit

* **New Features**
  * Added v2 adapter infrastructure for converting and validating route data with `strict`/`prune` adaptation modes, including PaRoutes adapter support.
  * Introduced v2 models: `Task`, `Target`, `Candidate`, `Benchmark`, and `FailureRecord` with constraint validation.
  * Added `ErrorCode` type for structured error handling.

* **Bug Fixes**
  * Corrected `InChIKeyStr` type capitalization with backward-compatible deprecation alias.

* **Documentation**
  * Updated schema documentation for type corrections.

* **Tests**
  * Added comprehensive adapter contract and model validation test coverage.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ischemist/project-procrustes/pull/102?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<h3>Greptile Summary</h3>

This PR introduces the v2 adapter layer for PaRoutes: protocol primitives (`Adapter`, `RawRouteEntry`, `AdaptMode`), five new Pydantic models (`Target`, `Task`, `Benchmark`, `Candidate`, `FailureRecord`), and a `PaRoutesAdapter` that converts raw target-keyed PaRoutes payloads into v2 `Route` objects with patent extraction, SMILES canonicalisation, and strict/prune adaptation modes. It also corrects the `InChIKeyStr` capitalisation with a backward-compatible alias.

- **Adapter core** (`paroutes.py`): `iter_raw_routes` detects single-route vs. target-keyed dict payloads and validates each root; `cast` runs patent-ID collection (`_get_patent_ids`) then recursive molecule construction (`_build_molecule`), rejecting cycles, mixed patents, empty reactions, and target SMILES mismatches.
- **Models** (`task.py`, `candidates.py`): `Target` cross-validates `smiles`/`inchikey` at construction; `Candidate` enforces exactly one of `route` or `failure`; `Task` enforces key\u2013id alignment across `targets`.
- **Tests**: a reusable `AdapterContractSuite` for future adapters, regression tests with stable route signatures against `paroutes.json.gz`, and per-behaviour contract tests covering all documented rejection paths.

<h3>Confidence Score: 4/5</h3>

Safe to merge with awareness of the known _get_patent_ids / _build_molecule in_stock traversal divergence noted in the prior review round.

The adapter's two-pass design contains a structural asymmetry where _get_patent_ids does not mirror _build_molecule's in_stock early-exit, meaning patent IDs can be collected from subtrees that route construction silently discards. Real PaRoutes data has in_stock molecules as leaves so regression tests pass, but the invariant is unenforced in code. Models and tests are otherwise sound.

src/retrocast/v2/adapters/paroutes.py — specifically the relationship between _get_patent_ids and _build_molecule traversal rules.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/retrocast/v2/adapters/paroutes.py | Core adapter converting raw PaRoutes JSON to v2 Route objects; contains the known _get_patent_ids / _build_molecule in_stock traversal divergence and a misleading error code when root SMILES is invalid in prune mode. |
| src/retrocast/v2/adapters/base.py | Defines the Adapter Protocol and RawRouteEntry dataclass; clean, minimal, and correct. |
| src/retrocast/v2/models/candidates.py | Adds FailureRecord and Candidate models with mutual-exclusivity validator; clean and well-tested. |
| src/retrocast/v2/models/task.py | Adds Target (with smiles/inchikey cross-validation), Task, TaskConstraints, and Benchmark models; correct and well-tested. |
| tests/v2/adapters/test_paroutes.py | Comprehensive contract, regression, and edge-case test suite covering cycles, mixed patents, empty reactions, prune mode, and stable route signatures against a real fixture. |
| tests/v2/adapters/base.py | Reusable adapter contract suite with dataclass-based case fixtures; clean design for future adapter implementations. |
| src/retrocast/typing.py | Corrects InChIKeyStr capitalisation, adds backward-compatible InchiKeyStr alias, and introduces ErrorCode NewType. |
| src/retrocast/v2/models/route.py | Minimal change: updates import from deprecated InchiKeyStr to InChIKeyStr. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Caller
    participant PaRoutesAdapter
    participant _validate_route_root
    participant _get_patent_ids
    participant _build_molecule

    Caller->>PaRoutesAdapter: iter_raw_routes(raw_payload)
    PaRoutesAdapter->>_validate_route_root: raw_dict (per target key)
    _validate_route_root-->>PaRoutesAdapter: PaRoutesMoleculeInput
    PaRoutesAdapter-->>Caller: RawRouteEntry(payload, source_key, source_order)

    Caller->>PaRoutesAdapter: cast(raw_route, target, mode)
    PaRoutesAdapter->>_validate_route_root: raw_route
    PaRoutesAdapter->>_get_patent_ids: validated root
    note over _get_patent_ids: cycle detection, patent ID via children[0]
    _get_patent_ids-->>PaRoutesAdapter: set[str] exactly 1
    PaRoutesAdapter->>_build_molecule: "validated root, visited={}"
    note over _build_molecule: cycle detection, in_stock early-exit, prune
    _build_molecule-->>PaRoutesAdapter: Molecule or None
    PaRoutesAdapter->>PaRoutesAdapter: SMILES match vs target
    PaRoutesAdapter-->>Caller: "Route(target, annotations={patent_id})"
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `src/retrocast/v2/adapters/paroutes.py`, line 338-351 ([link](https://github.com/ischemist/project-procrustes/blob/2f0702d015d3a8dbf0580b23726b029c9fec087c/src/retrocast/v2/adapters/paroutes.py#L338-L351)) 

   <a href="#">P1</a> **`_get_patent_ids` diverges from `_build_molecule` traversal rules**

   `_get_patent_ids` descends into every child reaction of every molecule node unconditionally, but `_build_molecule` applies two early-exit rules that are absent here:

   1. **`in_stock` molecules are treated as leaves** — `_build_molecule` returns immediately when `raw_mol_node.in_stock` is true, but `_get_patent_ids` still recurses into their children and collects patent IDs from synthesis routes that will never be included in the built route.
   2. **Only the first child reaction is used** — `_build_molecule` warns and uses only `children[0]`, but `_get_patent_ids` iterates over all reactions in the `for reaction_node in node.children` loop.

   The result is that `_get_patent_ids` can collect patent IDs from parts of the tree that `_build_molecule` will silently discard, causing a false `adapter.multiple_patents` rejection for a route that would in practice be built from a single patent's data.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/retrocast/v2/adapters/paroutes.py
   Line: 338-351

   Comment:
   **`_get_patent_ids` diverges from `_build_molecule` traversal rules**

   `_get_patent_ids` descends into every child reaction of every molecule node unconditionally, but `_build_molecule` applies two early-exit rules that are absent here:

   1. **`in_stock` molecules are treated as leaves** — `_build_molecule` returns immediately when `raw_mol_node.in_stock` is true, but `_get_patent_ids` still recurses into their children and collects patent IDs from synthesis routes that will never be included in the built route.
   2. **Only the first child reaction is used** — `_build_molecule` warns and uses only `children[0]`, but `_get_patent_ids` iterates over all reactions in the `for reaction_node in node.children` loop.

   The result is that `_get_patent_ids` can collect patent IDs from parts of the tree that `_build_molecule` will silently discard, causing a false `adapter.multiple_patents` rejection for a route that would in practice be built from a single patent's data.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 1 code review issue. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 1
src/retrocast/v2/adapters/paroutes.py:344-352
**Invalid root SMILES in prune mode surfaces as wrong error code**

When the root molecule's `smiles` string fails `canonicalize_smiles`, `_get_patent_ids` returns an empty set in prune mode (line 350–351). Control returns to `cast`, which immediately raises `adapter.patent_id_missing`. The caller sees a misleading "no patent id" error instead of anything pointing to an invalid target SMILES. Because `_build_molecule` is never reached, `adapter.target_pruned` is also not raised. An explicit guard at the top of `cast` — canonicalizing the root's SMILES before calling `_get_patent_ids` — would surface the real failure with an accurate error code in both modes.


`````

</details>

<sub>Reviews (2): Last reviewed commit: ["Address v2 adapter review findings"](https://github.com/ischemist/project-procrustes/commit/66e166e823f07aba88627d17a44f8f81db46bbb2) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=34292281)</sub>

### CODE REVIEW

File: src/retrocast/v2/adapters/paroutes.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+        raw_payload: Any,
+        *,
+        source_key: str | None = None,
+    ) -> Iterator[RawRouteEntry]:
+        if not isinstance(raw_payload, Mapping):
+            target_id = source_key or "<unknown>"
+            raise adapter_schema_error("paroutes", target_id, "expected route root or mapping of target ids to routes")
+
+        is_route_root = raw_payload.get("type") == "mol" or "smiles" in raw_payload or "children" in raw_payload
+        if is_route_root:
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply src/retrocast/v2/adapters/paroutes.py:153]: um, why would you call target "smiles" or "children" lmao, that's on you buddy

File: src/retrocast/v2/adapters/paroutes.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+            raise adapter_cycle_error("paroutes", node.smiles)
+
+        patent_ids: set[str] = set()
+        new_visited = visited | {node.smiles}
+        for reaction_node in node.children:
+            reaction_node = self._require_reaction_node(reaction_node, role="molecule child")
+            patent_ids.add(reaction_node.metadata.id.split(";")[0])
+            for reactant_node in reaction_node.children:
+                patent_ids.update(self._get_patent_ids(reactant_node, visited=new_visited))
+        return patent_ids
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply src/retrocast/v2/adapters/paroutes.py]: good point

File: src/retrocast/v2/adapters/paroutes.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+                reactants.append(reactant)
+        if not reactants:
+            if mode == "prune":
+                return None
+            raise AdapterLogicError(
+                "PaRoutes reaction has no reactants",
+                code="adapter.reaction_empty",
+                context={"adapter": "paroutes", "smiles": canon_smiles},
+            )
+
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply src/retrocast/v2/adapters/paroutes.py:297]: if you call your target "smiles" that's on you
